akka-stream/src/main/scala/akka/stream/impl/io/TcpStages.scala

```diff
@@ -0,0 +1,320 @@
+            failStage(ex)
+          case c: Connected ⇒
+            push(out, connectionFor(c, sender))
+          case Unbind ⇒
+            if (!isClosed(out) && (listener ne null)) listener ! Unbind
+          case Unbound ⇒ // If we're unbound then just shut down
+            completeStage()
+          case Terminated(ref) if ref == listener ⇒
+            failStage(new IllegalStateException("IO Listener actor terminated unexpectedly"))
+        }
+      }
+
+      setHandler(out, new OutHandler {
+        override def onPull(): Unit = {
+          // Ignore if still binding
+          if (listener ne null) listener ! ResumeAccepting(1)
+        }
+
+        override def onDownstreamFinish(): Unit = tryUnbind()
+      })
+
+      private def connectionFor(connected: Connected, connection: ActorRef): StreamTcp.IncomingConnection = {
+        val tcpFlow =
+          Flow.fromGraph(new IncomingConnectionStage(connection, connected.remoteAddress, halfClose))
+            .via(new Detacher[ByteString]) // must read ahead for proper completions
+
+        // FIXME: Previous code was wrong, must add new tests
+        val handler = idleTimeout match {
+          case d: FiniteDuration ⇒ tcpFlow.join(BidiFlow.bidirectionalIdleTimeout[ByteString, ByteString](d))
+          case _                 ⇒ tcpFlow
+        }
+
+        StreamTcp.IncomingConnection(
+          connected.localAddress,
+          connected.remoteAddress,
+          handler)
+      }
+
+      private def tryUnbind(): Unit = {
+        if (listener ne null) listener ! Unbind
+        else completeStage()
+      }
+
+      override def postStop(): Unit = {
+        unbindPromise.trySuccess(())
+        bindingPromise.tryFailure(new NoSuchElementException("Binding was unbound before it was completely finished"))
+      }
+    }
+
+    (logic, bindingPromise.future)
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+private[stream] object TcpConnectionStage {
+  case object WriteAck extends Tcp.Event
+
+  trait TcpRole {
+    def halfClose: Boolean
+  }
+  case class Outbound(
+    manager: ActorRef,
+    connectCmd: Connect,
+    localAddressPromise: Promise[InetSocketAddress],
+    halfClose: Boolean) extends TcpRole
+  case class Inbound(connection: ActorRef, halfClose: Boolean) extends TcpRole
+
+  /*
+   * This is a *non-deatched* design, i.e. this does not prefetch itself any of the inputs. It relies on downstream
+   * stages to provide the necessary prefetch on `bytesOut` and the framework to do the proper prefetch in the buffer
+   * backing `bytesIn`. If prefetch on `bytesOut` is required (i.e. user stages cannot be trusted) then it is better
+   * to attach an extra, fused buffer to the end of this flow. Keeping this stage non-detached makes it much simpler and
+   * easier to maintain and understand.
+   */
+  class TcpStreamLogic(val shape: FlowShape[ByteString, ByteString], val role: TcpRole) extends GraphStageLogic(shape) {
```

Comment:
does this need its own field etc.? (i.e. why not call `_shape` and just pass to `super()`?)